clients/command/src/main/java/org/eclipse/hono/client/command/CommandRouterCommandConsumerFactory.java

```diff
@@ -288,4 +354,42 @@ private Future<Void> removeCommandConsumer(
                     }
                 });
     }
+
+    private Future<Void> handleClosingConsumers(final SpanContext onReleaseSpanContext) {
+        if (pendingConsumersToClose.size() >= batchSize) {
+            return removeCommandConsumers(pendingConsumersToClose.size(), onReleaseSpanContext);
+        } else {
+            // start the timer if not started
+            if (batchTimerId == -1) {
+                final Vertx vertx = Vertx.currentContext().owner();
+                if (vertx != null) {
+                    batchTimerId = vertx.setTimer(batchMaxTimeout, aLong -> {
+                        final Span span = TracingHelper
+                                .buildSpan(tracer, null, "on consumers timer", References.CHILD_OF)
+                                .ignoreActiveSpan()
+                                .withTag(Tags.COMPONENT.getKey(), "CommandConsumerTimer")
+                                .start();
+                        removeCommandConsumers(pendingConsumersToClose.size(), span.context())
+                                .onComplete(v -> span.finish());
+                        // mark the timer as finished although the batch operation is still ongoing
+                        batchTimerId = -1;
+                    });
+                }
+            }
+            return Future.succeededFuture();
+        }
+    }
+
+    private Future<Void> removeCommandConsumers(final int count, final SpanContext spanContext) {
+        if (count > 0) {
+            final List<CommandRouterDeviceInfo> toSend = new ArrayList<>();
+            for (int i = 0; i < count; i++) {
+                final CommandRouterDeviceInfo crdi = pendingConsumersToClose.remove(0);
```

Comment:
It could be that first the vertx timer for a `removeCommandConsumers` invocation gets created, and then, before the timer has elapsed, a single `consumer.close()` invocation gets done (doing a `pendingConsumersToClose.removeIf()`).

That means the subsequent `removeCommandConsumers` invocation (triggered by the timer) may find the `pendingConsumersToClose` size to be less than `count`.

Reply:
Sorry, got that wrong, `count` still being given the current `pendingConsumersToClose` size.